go/store/datas/puller.go

```diff
@@ -0,0 +1,294 @@
+	"sync"
+
+	"github.com/liquidata-inc/dolt/go/store/atomicerr"
+	"github.com/liquidata-inc/dolt/go/store/chunks"
+	"github.com/liquidata-inc/dolt/go/store/hash"
+	"github.com/liquidata-inc/dolt/go/store/nbs"
+	"github.com/liquidata-inc/dolt/go/store/types"
+)
+
+const (
+	maxChunkWorkers = 2
+)
+
+// FilledWriters store CmpChunkTableWriter that have been filled and are ready to be flushed.  In the future will likely
+// add the md5 of the data to this structure to be used to verify table upload calls.
+type FilledWriters struct {
+	wr *nbs.CmpChunkTableWriter
+}
+
+// CmpChnkAndRefs holds a CompressedChunk and all of it's references
+type CmpChnkAndRefs struct {
+	cmpChnk nbs.CompressedChunk
+	refs    map[hash.Hash]bool
+}
+
+// Puller is used to sync data between to Databases
+type Puller struct {
+	fmt *types.NomsBinFormat
+
+	srcDB         Database
+	sinkDB        Database
+	rootChunkHash hash.Hash
+	downloaded    hash.HashSet
+
+	wr          *nbs.CmpChunkTableWriter
+	tempDir     string
+	chunksPerTF int
+}
+
+// NewPuller creates a new Puller instance to do the syncing.  If a nil puller is returned without error that means
+// that there is nothing to pull and the sinkDB is already up to date.
+func NewPuller(ctx context.Context, tempDir string, chunksPerTF int, srcDB, sinkDB Database, rootChunkHash hash.Hash) (*Puller, error) {
+	// Sanity Check
+	exists, err := srcDB.chunkStore().Has(ctx, rootChunkHash)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, errors.New("not found")
+	}
+
+	exists, err = sinkDB.chunkStore().Has(ctx, rootChunkHash)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if exists {
+		return nil, nil // already up to date
+	}
+
+	if srcDB.chunkStore().Version() != sinkDB.chunkStore().Version() {
+		return nil, fmt.Errorf("cannot pull from src to sink; src version is %v and sink version is %v", srcDB.chunkStore().Version(), sinkDB.chunkStore().Version())
+	}
+
+	wr, err := nbs.NewCmpChunkTableWriter()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Puller{
+		fmt:           srcDB.Format(),
+		srcDB:         srcDB,
+		sinkDB:        sinkDB,
+		rootChunkHash: rootChunkHash,
+		downloaded:    hash.HashSet{},
+		tempDir:       tempDir,
+		wr:            wr,
+		chunksPerTF:   chunksPerTF,
+	}, nil
+}
+
+func (p *Puller) processCompletedTables(ctx context.Context, ae *atomicerr.AtomicError, completedTables <-chan FilledWriters) {
+	type tempTblFile struct {
+		id        string
+		path      string
+		numChunks int
+	}
+
+	var tblFiles []tempTblFile
+
+	var err error
+	for tblFile := range completedTables {
+		if err != nil {
+			continue // drain
+		}
+
+		var id string
+		id, err = tblFile.wr.Finish()
+
+		if ae.SetIfError(err) {
+			continue
+		}
+
+		path := filepath.Join(p.tempDir, id)
+		err = tblFile.wr.FlushToFile(path)
+
+		if ae.SetIfError(err) {
+			continue
+		}
+
+		tblFiles = append(tblFiles, tempTblFile{id, path, tblFile.wr.Size()})
+	}
+
+	for i := len(tblFiles) - 1; i >= 0; i-- {
```

Comment:
Should this block run even if the above block errored out?

Reply:
fixed